docs/DafnyRef/Topics.md

```diff
@@ -34,6 +34,20 @@ const m := set x | n in {-1, 0, 1, 2} && f(n) >= 1
 const m := set x | n in {-1, 0, 1, 2} && g(n) >= 1
 ```
 
+A special case to be aware of: if the type in the comprehension differs from the collection,
+but both are compilable, the type of the comprehension will be fully tested, which could be viewed as a performance problem if only one specific trait needed to be tested. For example:
+
+```dafny
+const s: set<nat> := {1, 2, 3}
+type NonNegative = x | x > 0 witness 1
+
+// not only x > 0 will be tested, but also x >= 0 (from nat)
```

Comment:
Did you mean `(from NonNegative)` ?

Reply:
You're right. I just added an optimization in the compiler so that if it reaches the original collection's type, it stops emitting constraints. Thus I also changed the example in this documentation to reflect a case when the compiler cannot arrive to the same type of the collection, and emits all constraints until `int`. Thus, the trick is still valid afterwards.